libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+    response.opcode = FTP_OP::Nack;
+    response.data[0] = static_cast<uint8_t>(error);
+    response.size = 1;
+
+    // FIXME: errno's are not thread-local as they should be on ChibiOS
+    if (error == FTP_ERROR::FailErrno) {
+        // translate the errno's that we have useful messages for
+        switch (errno) {
+            case EEXIST:
+                response.data[0] = static_cast<uint8_t>(FTP_ERROR::FileExists);
+                break;
+            case ENOENT:
+                response.data[0] = static_cast<uint8_t>(FTP_ERROR::FileNotFound);
+                break;
+            default:
+                response.data[1] = static_cast<uint8_t>(errno);
+                response.size = 2;
+                break;
+        }
+    }
+}
+
+void GCS_MAVLINK::ftp_worker(void) {
+    pending_ftp request;
+    pending_ftp reply = {};
+    reply.session = -1; // flag the reply as invalid for any reuse
+
+    while (true) {
+        while (!ftp.requests->pop(request)) {
+            // nothing to handle delay ourselves a bit then check again, ideally we'd use conditional waits here
+            hal.scheduler->delay(10);
+        }
+
+        // if it's a rerequest and we still have the last response then send it
+        if ((request.sysid == reply.sysid) && (request.compid = reply.compid) &&
+            (request.session == reply.session) && (request.seq_number + 1 == reply.seq_number)) {
+            while (!ftp.replies->push(reply)) {
+                hal.scheduler->delay(10);
+            }
+            continue;
+        }
+
+        // setup the response
+        memset(&reply, 0, sizeof(pending_ftp));
+        reply.req_opcode = request.opcode;
+        reply.session = request.session;
+        reply.seq_number = request.seq_number + 1;
+        reply.chan = request.chan;
+        reply.sysid = request.sysid;
+        reply.compid = request.compid;
+
+        // sanity check the request size
+        if (request.size > sizeof(request.data)) {
+            ftp_error(reply, FTP_ERROR::InvalidDataSize);
+            while (!ftp.replies->push(reply)) {
+                hal.scheduler->delay(10);
+            }
+            continue;
+        }
+
+        // reject anything not belonging to our current session, unless we are terminating a session, then just kill it and move on
+        if (((ftp.current_session >= 0) && (request.session != ftp.current_session))
+            && ((request.opcode == FTP_OP::TerminateSession) || (request.opcode == FTP_OP::ResetSessions))) {
+            reply.opcode = FTP_OP::Ack;
+        } else {
+            // dispatch the command as needed
+            switch (request.opcode) {
+                case FTP_OP::None:
+                    reply.opcode = FTP_OP::Ack;
+                    break;
+                case FTP_OP::TerminateSession:
+                case FTP_OP::ResetSessions:
+                    // we already handled this, just listed for completion
```

Comment:
Should this be 

```suggestion
                    // we already handled this, just listed for completeness
```

?